test/text_region/test_text_columns.py

```diff
@@ -0,0 +1,193 @@
+    with cols_3:
+        pdf.set_font("Times", "", 14)
+        cols_3.write(text=LOREM_IPSUM[:300])
+    assert_pdf_equal(pdf, HERE / "tcols_balance.pdf", tmp_path)
+
+
+def test_tcols_charwrap(tmp_path):
+    pdf = FPDF()
+    pdf.add_page()
+    pdf.set_font("courier", "", 16)
+    col = pdf.text_columns(l_margin=50, r_margin=50)
+    # wrapmode on paragraph
+    with col.paragraph(wrapmode="CHAR", bottom_margin=pdf.font_size) as par:
+        par.write(text=LOREM_IPSUM[:500])
+    col.render()
+    # wrapmode on column
+    with pdf.text_columns(
+        # align="J",
+        l_margin=50,
+        r_margin=50,
+        wrapmode="CHAR",
+    ) as col:
+        with col.paragraph() as par:
+            par.write(text=LOREM_IPSUM[500:1000])
+    assert_pdf_equal(pdf, HERE / "tcols_charwrap.pdf", tmp_path)
+
+
+def test_tcols_no_font():
+    pdf = FPDF()
+    pdf.add_page()
+    with pytest.raises(FPDFException) as error:
+        col = pdf.text_columns()
+        col.write("something")
+    expected_msg = "No font set, you need to call set_font() beforehand"
+    assert str(error.value) == expected_msg
+    with pytest.raises(FPDFException) as error:
+        col.ln()
+    expected_msg = "No font set, you need to call set_font() beforehand"
+    assert str(error.value) == expected_msg
+
+
+def test_tcols_bad_uses():
+    pdf = FPDF()
+    pdf.add_page()
+    pdf.set_font("courier", "", 16)
+    col = pdf.text_columns()
+    # recursive text region context
+    with col:
+        col.write("something")
+        with pytest.raises(FPDFException) as error:
+            with col:
+                pass
+    expected_msg = "Unable to enter the same TextColumns context recursively."
+    assert str(error.value) == expected_msg
+    # recursive use of paragraph context
+    with col.paragraph() as par:
+        par.write("something")
+        with pytest.raises(FPDFException) as error:
+            col.paragraph()
+    expected_msg = "Unable to nest paragraphs."
+    assert str(error.value) == expected_msg
+    # writing to column while we have an explicit paragraph active
+    with col.paragraph() as par:
+        par.write("something")
+        with pytest.raises(FPDFException) as error:
+            col.write("else")
+    expected_msg = "Conflicts with active paragraph. Either close the current paragraph or write your text inside it."
+    assert str(error.value) == expected_msg
+    # ending a non-existent paragraph
+    with pytest.raises(FPDFException) as error:
+        col.end_paragraph()
+    expected_msg = "No active paragraph to end."
+    assert str(error.value) == expected_msg
+    # column with negative width
+    with pytest.raises(FPDFException) as error:
+        col = pdf.text_columns(l_margin=150, r_margin=150)
+    expected_msg = (
+        "TextColumns(): Right limit (60.00155555555551) lower than left limit (150)."
+    )
+    assert str(error.value) == expected_msg
+
+
+def xest_tcols_text_shaping(tmp_path):
```

Comment:
Mayeb you could use `@pytest.mark.skip(reason="...")` there?
https://docs.pytest.org/en/7.3.x/how-to/skipping.html

Reply:
Then I'd have to remember that syntax... 😛